plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,155 @@
+		m_free_indexes_stack = (uint16*)&m_buffer[END];
+		// Initialize head
+		m_buffer[0].data = NULL;
+		m_buffer[0].next_index = END;
+		m_buffer[0].prev_index = 65535;
+		// Stack top is 0, bottom is END - 2
+		m_free_indexes_stack_top = 0;
+		// Index 0 reserved for head
+		for (uint16 i = 0; i < END - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+	}
+
+	__forceinline bool Empty() {
+		return m_free_indexes_stack == 0;
+	}
+
+	__forceinline bool Full() {
+		return m_free_indexes_stack_top == END - 1;
+	}
+
+	// Insert the element in front of the list and return it's position in m_buffer
+	__forceinline uint16 InsertFront(const T data) {
+		if (Full()) {
+			Grow();
+		}
+		// Pop a free index from the stack
+		const uint16 free_index = m_free_indexes_stack[m_free_indexes_stack_top++];
+		m_buffer[free_index].data = data;
+		ListInsertFront(free_index);
+		return free_index;
+	}
+
+	__forceinline void Erase(const uint16 index) {
+		ListRemove(index);
+		m_free_indexes_stack[--m_free_indexes_stack_top] = index;
+	}
+
+	__forceinline void MoveFront(const uint16 index) {
+		if (m_buffer[0].next_index != index) {
+			ListRemove(index);
+			ListInsertFront(index);
+		}
+	}
+private:
+	uint16 m_free_indexes_stack_top;
+	// m_buffer indexes that are free to be used are stacked here [END - 1]
+	uint16* m_free_indexes_stack;
+
+	__forceinline void ListInsertFront(const uint16 index) {
+		// Update prev / next indexes to add m_buffer[index] to the chain
+		Element<T>& head = m_buffer[0];
+		m_buffer[index].prev_index = 0;
+		m_buffer[index].next_index = head.next_index;
+		if (head.next_index != END) {
+			m_buffer[head.next_index].prev_index = index;
+		}
+		head.next_index = index;
+	}
+
+	__forceinline void ListRemove(const uint16 index) {
+		// Update prev / next indexes to remove m_buffer[index] from the chain
+		const Element<T>& to_remove = m_buffer[index];
+		m_buffer[to_remove.prev_index].next_index = to_remove.next_index;
+		if (to_remove.next_index != END) {
+			m_buffer[to_remove.next_index].prev_index = to_remove.prev_index;
+		}
+	}
+
+	void Grow() {
+		if (END == 65535) {
+			throw std::runtime_error("List size maxed out at 65535 elements, cannot grow futhermore.");
+		}
+
+		const uint16 NEW_END = END >= 32768 ? 65535 : END * 2;
+
+		m_buffer = (Element<T>*)_aligned_realloc(m_buffer, NEW_END * sizeof(Element<T>) + (NEW_END - 1) * sizeof(uint16), 64);
+		uint16* new_free_indexes_stack = (uint16*)&m_buffer[NEW_END];
+		memcpy(new_free_indexes_stack, (uint16*)&m_buffer[END], (END - 1) * sizeof(uint16));
+		m_free_indexes_stack = new_free_indexes_stack;
+
+		// Replace the END marker with the updated one. This is the easiest way and does not require additional class members
+		for (uint16 i = m_buffer[0].next_index; i != END; i = m_buffer[i].next_index) {
+			if (m_buffer[i].next_index == END) {
+				m_buffer[i].next_index = NEW_END;
+				break;
+			}
+		}
```

Comment:
By using constant value for last next_index, the above piece of code (140-146) could be removed.